### PR TITLE
Migrate addon storyshots to TS

### PR DESCRIPTION
**Pull request by gaetanmaisse**
## What I did

 - Migrate `@storybook/addon-storyshots` and `@storybook/addon-storyshots-puppeteer` to TypeScript
 - Add types and classes to have a simpler DX when working on this addon
 - Refactor import/export to use named exports instead of default ones

Tracking Issue: https://github.com/storybooks/storybook/issues/5030

### CODE REVIEW

**Comment by gaetanmaisse** [comment]
@shilman @ndelangen should all stories from [storyshots-core](https://github.com/storybookjs/storybook/tree/741e69584eeb9a63c2e7a0c9e300a4d9537ee959/addons/storyshots/storyshots-core/stories) be migrated to TS? kept to JS? Duplicated to TS? 
I also read a bit about CSF format, maybe they should be converted to it?

File: addons/storyshots/storyshots-core/src/api/index.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/api/index.ts]: 👀

File: addons/storyshots/storyshots-core/src/api/index.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/api/index.ts]: Useful?

File: addons/storyshots/storyshots-core/src/api/index.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/api/index.ts]: Need to investigate more, but it looks hard to type

File: addons/storyshots/storyshots-core/src/api/index.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/api/index.ts]: `string`?

File: addons/storyshots/storyshots-core/src/api/index.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/api/index.ts]: `string`?

File: addons/storyshots/storyshots-core/src/frameworks/react-native/loader.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/frameworks/react-native/loader.ts]: To rename

File: addons/storyshots/storyshots-puppeteer/src/imageSnapshot.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-puppeteer/src/imageSnapshot.ts]: to rename according to https://github.com/storybookjs/storybook/pull/7674/files#diff-0b05142d970443de5c41e81ad14df803R7?

File: addons/storyshots/storyshots-core/src/api/StoryshotsOptions.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/api/StoryshotsOptions.ts]: @ndelangen @kroeder are there already existing types somewhere for `story` and `context`?

File: addons/storyshots/storyshots-puppeteer/src/imageSnapshot.ts
Comment (gaetanmaisse) [reply addons/storyshots/storyshots-puppeteer/src/imageSnapshot.ts]: Moreover it looks like it match https://github.com/storybookjs/storybook/pull/7674/files#diff-d72af9277db9c8a6cb1b3595750de8c4R9 which is not currently true

File: addons/storyshots/storyshots-core/package.json
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/package.json]: @emilio-martinez @kroeder @ndelangen what's the best practice about `@types` in SB codebase? I added them in `dependencies` because some of the types I exported in this addon are based on `jest` (and other libs) ones. So if a SB user want to use SB type it should install `@types/jest` too... 🤷‍♂

File: addons/storyshots/storyshots-puppeteer/package.json
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-puppeteer/package.json]: Same question as https://github.com/storybookjs/storybook/pull/7674/files#r312234521

File: addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts
Comment (gaetanmaisse) [inline addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts]: It's the way I found to have type-check according to `Loader` class, if anyone has a better proposition?

**Comment by shilman** [comment]
@gaetanmaisse @ndelangen Let's merge this after 5.2 is released please 🙏

**Comment by ndelangen** [comment]
hey @gaetanmaisse I merged in next, do you think you'd be able to finish this PR?

**Comment by gaetanmaisse** [comment]
@ndelangen I just come back from holidays I'm on it 😉

File: addons/storyshots/storyshots-core/src/api/index.ts
Comment (gaetanmaisse) [reply addons/storyshots/storyshots-core/src/api/index.ts]: Resolved

**Comment by gaetanmaisse** [comment]
A lot of types are still missing but I think I reach a good state to merge this PR and continue my work on this after it will be merged.

During reviews, a focus on my last commit would be great because I think this part is not tested during PR automatic tests (but I'm maybe wrong?) and I may have incorrectly understood how this puppeteer related addon works.

**Comment by ndelangen** [comment]
What's the difference between puppeteer & puppeteer-core?

**Comment by ndelangen** [comment]
So awesome you got this green 🎉

File: addons/storyshots/storyshots-core/src/api/StoryshotsOptions.ts
Comment (ndelangen) [reply addons/storyshots/storyshots-core/src/api/StoryshotsOptions.ts]: Yes, in lib/addons

**Comment by gaetanmaisse** [comment]
@ndelangen 
> What's the difference between puppeteer & puppeteer-core?

See commit message of https://github.com/storybookjs/storybook/pull/7674/commits/749c3944d2cd7ef8dfb8fd6837706e6e6d244e0a : 
>`puppeteer-core` package is a version of `puppeteer` that doesn't download Chromium by default.
It looks like chromium is not directly used here as it was only a `peerDependencies` or an `optionalDependencies`.
So I moved to `puppeteer-core` to be able to compile TS sources of this addon without downloading Chromium at every `yarn install`

**Comment by SimenB** [comment]
https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteer-vs-puppeteer-core

**Comment by ndelangen** [comment]
Awesome, merge when you think it's ready!

**Review by ndelangen — APPROVED** [review]

**Review by shilman — COMMENTED** [review]
More `any` types than I'd expect, but I'm assuming there's good reason for it?

File: addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts
Comment (gaetanmaisse) [reply addons/storyshots/storyshots-core/src/frameworks/angular/loader.ts]: Good catch! I fixed it.

**Comment by gaetanmaisse** [comment]
@shilman 
> More any types than I'd expect, but I'm assuming there's good reason for it?

A lot of them are related to `story` or `context` function arguments, I tried to use some types defined in `lib/addons` but at the end, I have some conflicting/incompatible types 😞 So what I have in mind is merging this PR to share the migration to everybody. Then continue to work to improve typings of `storyshot` addons and make a 2nd PR when it will be ready.

**Pull request merged**
